### ISSUE TITLE
regression related to fully qualified domain names

### ISSUE THREAD

**Issue by ananthakumaran** [benoitc/hackney#607]
```elixir
:recon_trace.calls(
  [{:idna, :to_ascii, :return_trace}, {:ssl, :connect, :return_trace}],
  100
)

{:ok, 200, _, _} = :hackney.request(:get, "https://www.example.com.", [], "", [])
```

## Hackney 1.13.0

```
10:57:46.644839 <0.92.0> idna:to_ascii("www.example.com.")

10:57:46.648138 <0.92.0> idna:to_ascii/1 --> "www.example.com"

10:57:46.669702 <0.92.0> ssl:connect("www.example.com", 443, [{ciphers, ..},
 {versions,['tlsv1.2','tlsv1.1',tlsv1,sslv3]},
 {honor_cipher_order,true},
 {reuse_sessions,true},
 {secure_renegotiate,true},
 {packet,raw},
 {active,false},
 binary,
 {verify,verify_peer},
 {depth,99},
 {cacerts, ..},
 {partial_chain,#Fun<hackney_connect.0.3608892>},
 {verify_fun,{fun ssl_verify_hostname:verify_fun/3,
              [{check_hostname,"www.example.com"}]}}], 8000)

10:57:47.446935 <0.92.0> ssl:connect/4 --> {ok,
                                            {sslsocket,
                                             {gen_tcp,#Port<0.5>,
                                              tls_connection,undefined},
                                             [<0.196.0>,<0.195.0>]}}
```

idna:to_ascii used to remove the trailing dot and the ssl module will always get the hostname without the trailing dot

## Hackney 1.15.2

```
10:58:58.472589 <0.92.0> ssl:connect("www.example.com.", 443, [{ciphers, ..},
 {versions,['tlsv1.2','tlsv1.1',tlsv1,sslv3]},
 {reuse_sessions,true},
 {secure_renegotiate,true},
 {packet,raw},
 {active,false},
 binary,
 {verify,verify_peer},
 {depth,99},
 {cacerts, ..},
 {partial_chain,#Fun<hackney_connect.0.8369309>},
 {verify_fun,{fun ssl_verify_hostname:verify_fun/3,
              [{check_hostname,"www.example.com."}]}}], 8000)

10:58:58.997 [info]  ['TLS', 32, 'client', 58, 32, 73, 110, 32, 115, 116, 97, 116, 101, 32, 'hello', 32, 'received SERVER ALERT: Warning - Unrecognised Name', 10]

10:58:59.259085 <0.92.0> ssl:connect/4 --> {ok,
                                            {sslsocket,
                                             {gen_tcp,#Port<0.525>,
                                              tls_connection,undefined},
                                             [<0.345.0>,<0.344.0>]}}

```

the call to idna:to_ascii has been removed and the hostname is passed with trailing dot to ssl module. ssl module use the hostname [with the trailing dot](https://github.com/erlang/otp/blob/eaaf721457bac35c18980c5ad48ba4373f2b4145/lib/ssl/src/ssl.erl#L2528-L2531) as the sni server name

```erlang
server_name_indication_default(Host) when is_list(Host) ->
    Host;
server_name_indication_default(_) ->
    undefined.
```

As per [RFC 6066 §3](https://tools.ietf.org/html/rfc6066.html#section-3)

```
The hostname is represented as a byte string using ASCII encoding without a trailing dot.
```

The client should send sni server name without trailing dot. Some of the ssl servers are ok with trailing dots, but some fail and return

```
11:21:14.572 [info]  ['TLS', 32, 'client', 58, 32, 73, 110, 32, 115, 116, 97, 116, 101, 32, 'hello', 32, 'received SERVER ALERT: Fatal - Unexpected Message', 10]
{:error,
 {:tls_alert,
  {:unexpected_message, 'received CLIENT ALERT: Fatal - Unexpected Message'}}}
```

Honestly I don't know where this should be fixed, whether in hackney or in ssl module.

**Comment by ananthakumaran** [comment]
erlang 22.3 might come with a fix for this issue https://bugs.erlang.org/browse/ERL-1135

**Comment by benoitc** [comment]
fixed via 6b6337299c3f106f1b431d00c824239186fa94f6 . Thanks for the report.
